app/components/Form/ProjectQuarterlyReportForm.tsx

```diff
@@ -205,10 +206,56 @@ const ProjectQuarterlyReportForm: React.FC<Props> = (props) => {
     return [filteredReports, nextIndex];
   }, [projectRevision.projectQuarterlyReportFormChanges]);
 
+  const handleUndo = async () => {
+    let committedQuarterlyReports = [];
+
+    projectRevision.projectQuarterlyReportFormChanges.edges.forEach(
+      ({ node }) => {
+        if (
+          node.formChangeByPreviousFormChangeId?.changeStatus === "committed"
+        ) {
+          committedQuarterlyReports.push(node);
+        }
+        if (node.changeStatus === "pending" || node.changeStatus === "staged") {
+          deleteQuarterlyReport(node.id, "ARCHIVE");
+        }
+      }
+    );
+
+    const completedPromises: Promise<void>[] = [];
+    committedQuarterlyReports.forEach((node) => {
+      const undoneFormData = node.formChangeByPreviousFormChangeId.newFormData;
+      const promise = new Promise<void>((resolve, reject) => {
+        applyUpdateFormChangeMutation({
+          variables: {
+            input: {
+              id: node.id,
+              formChangePatch: {
+                changeStatus: "pending",
+                newFormData: undoneFormData,
+                operation: "UPDATE",
+              },
+            },
+          },
+          debounceKey: node.id,
+          onCompleted: () => resolve(),
+          onError: reject,
+        });
+      });
+      completedPromises.push(promise);
+    });
+    try {
+      await Promise.all(completedPromises);
+    } catch (e) {
+      // the failing mutation will display an error message and send the error to sentry
+    }
+  };
```

Comment:
There is a lot of code duplication going on between all of our front-end files. Can we create a database function instead that receives an array of `formChange.rowId`, undoes them if needed, and returns the corresponding form_change records?

Something like:

```sql
create function cif.undo_form_changes(form_change_ids int[])
returns setof cif.form_change
```

Reply:
Sure thing, I'm working on that with the help of Dylan.